### ISSUE TITLE
🐛 [firebase_auth] Link auth provider credentials to a user account with Apple credentials

### ISSUE THREAD

**Issue by jeanmatthieud** [firebase/flutterfire#9442]
## Bug report
I used the new `AppleProvider` [as documented](https://firebase.google.com/docs/auth/flutter/federated-auth#apple)

I want to be able to [link the credentials the another account](https://firebase.google.com/docs/auth/flutter/account-linking#link_auth_provider_credentials_to_a_user_account).

I should be able to do that with:
```dart
final appleProvider = AppleAuthProvider();
UserCredential? authResult;
if (kIsWeb) {
  authResult = await FirebaseAuth.instance.signInWithPopup(appleProvider);
} else {
  authResult = await FirebaseAuth.instance.signInWithAuthProvider(appleProvider);
}
FirebaseAuth.instance.currentUser!.linkWithCredential(authResult.credentials);
```

Or create some credentials with:
```dart
final credential = AppleAuthProvider.credential(authResult.credentials.token);
```

As it with required to reauthenticate the user before account deletion (for example):
```dart
await FirebaseAuth.instance.currentUser!.reauthenticateWithCredential(
  AppleAuthProvider.credential(token),
);
```

But, the `token` is empty (with Android and iOs).
**Without it, I cannot link the anonymous user to the Apple credentials, or delete the user account.**

---

## Additional context

- Tested with **Android 31** and **iOs 15.5**
- firebase_auth: ^3.7.0
- Not tested on Web

---

### Flutter doctor

<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.0.5, on Microsoft Windows [version 10.0.22000.856], locale fr-FR)
[!] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.0.6)
[√] Android Studio (version 2021.2)
[√] VS Code, 64-bit edition (version 1.69.2)
[√] Connected device (4 available)
[√] HTTP Host Availability
```

</details>

---

### Flutter dependencies

<details><summary>Click To Expand</summary>

```
Dart SDK 2.17.6
Flutter SDK 3.0.5
[snip] 0.0.0+1

dependencies:
- animations 2.0.4 [flutter]
- cloud_firestore 3.4.6 [cloud_firestore_platform_interface cloud_firestore_web collection firebase_core firebase_core_platform_interface flutter meta]
- cupertino_icons 1.0.5
- device_info_plus 4.1.2 [flutter device_info_plus_platform_interface device_info_plus_macos device_info_plus_linux device_info_plus_web device_info_plus_windows]
- dio 4.0.6 [http_parser path]
- duration 3.0.11
- firebase_analytics 9.3.3 [firebase_analytics_platform_interface firebase_analytics_web firebase_core firebase_core_platform_interface flutter]
- firebase_auth 3.7.0 [firebase_auth_platform_interface firebase_auth_web firebase_core firebase_core_platform_interface flutter meta]
- firebase_core 1.21.1 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_crashlytics 2.8.9 [firebase_core firebase_core_platform_interface firebase_crashlytics_platform_interface flutter stack_trace]
- firebase_in_app_messaging 0.6.0+23 [firebase_core firebase_core_platform_interface firebase_in_app_messaging_platform_interface flutter meta]
- firebase_messaging 13.0.0 [firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web flutter meta]
- firebase_remote_config 2.0.16 [firebase_core firebase_core_platform_interface firebase_remote_config_platform_interface firebase_remote_config_web flutter]
- floor 1.3.0 [floor_annotation flutter meta path sqflite sqflite_common_ffi]
- flutter 0.0.0 [characters collection material_color_utilities meta vector_math sky_engine]
- flutter_google_places 0.3.0 [flutter google_api_headers google_maps_webservice http rxdart]
- flutter_keyboard_visibility 5.3.0 [meta flutter_keyboard_visibility_platform_interface flutter_keyboard_visibility_web flutter]
- flutter_linkify 5.0.2 [flutter linkify]
- flutter_signin_button 2.0.0 [flutter font_awesome_flutter]
- flutter_svg 1.1.4 [flutter meta path_drawing vector_math xml]
- google_maps_flutter 2.1.12 [flutter google_maps_flutter_android google_maps_flutter_ios google_maps_flutter_platform_interface]
- google_maps_webservice 0.0.20-nullsafety.5 [http meta json_annotation]
- google_mobile_ads 2.0.1 [meta flutter visibility_detector]
- google_sign_in 5.4.1 [flutter google_sign_in_android google_sign_in_ios google_sign_in_platform_interface google_sign_in_web]
- in_app_purchase 3.0.7 [flutter in_app_purchase_android in_app_purchase_platform_interface in_app_purchase_storekit]
- in_app_review 2.0.4 [flutter in_app_review_platform_interface]
- intl 0.17.0 [clock path]
- intl_phone_number_input 0.7.0+2 [flutter libphonenumber_plugin equatable collection]
- json_annotation 4.6.0 [meta]
- lottie 1.4.2 [archive flutter path vector_math]
- nonce 1.2.0
- package_info_plus 1.4.3+1 [flutter package_info_plus_platform_interface package_info_plus_linux package_info_plus_macos package_info_plus_windows package_info_plus_web]
- permission_handler 10.0.0 [flutter meta permission_handler_android permission_handler_apple permission_handler_windows permission_handler_platform_interface]
- photo_view 0.14.0 [flutter]
- provider 6.0.3 [collection flutter nested]
- quick_actions 0.6.0+11 [flutter quick_actions_android quick_actions_ios quick_actions_platform_interface]
- shared_preferences 2.0.15 [flutter shared_preferences_android shared_preferences_ios shared_preferences_linux shared_preferences_macos shared_preferences_platform_interface shared_preferences_web shared_preferences_windows]
- shimmer 2.0.0 [flutter]
- sign_in_with_apple 4.1.0 [flutter meta sign_in_with_apple_platform_interface sign_in_with_apple_web]
- sqflite 2.0.3+1 [flutter sqflite_common path]
- url_launcher 6.1.5 [flutter url_launcher_android url_launcher_ios url_launcher_linux url_launcher_macos url_launcher_platform_interface url_launcher_web url_launcher_windows]

dev dependencies:
- build_runner 2.2.0 [args async analyzer build build_config build_daemon build_resolvers build_runner_core code_builder collection crypto dart_style frontend_server_client glob graphs http_multi_server io js logging meta mime package_config path pool pub_semver pubspec_parse shelf shelf_web_socket stack_trace stream_transform timing watcher web_socket_channel yaml]
- floor_generator 1.3.0 [analyzer build build_config code_builder collection floor_annotation meta source_gen strings]
- flutter_launcher_icons 0.10.0 [args checked_yaml cli_util image json_annotation path yaml]
- flutter_native_splash 2.2.6 [args flutter flutter_web_plugins image js lint meta path universal_io xml yaml]
- json_serializable 6.3.1 [analyzer async build build_config collection json_annotation meta path pub_semver pubspec_parse source_gen source_helper]

dependency overrides:
- libphonenumber 2.0.2 [flutter meta]

transitive dependencies:
- _fe_analyzer_shared 43.0.0 [meta]
- analyzer 4.3.1 [_fe_analyzer_shared collection convert crypto glob meta package_config path pub_semver source_span watcher yaml]
- archive 3.3.1 [crypto path]
- args 2.3.1
- async 2.8.2 [collection meta]
- boolean_selector 2.1.0 [source_span string_scanner]
- build 2.3.0 [analyzer async convert crypto glob logging meta path]
- build_config 1.1.0 [checked_yaml json_annotation path pubspec_parse yaml]
- build_daemon 3.1.0 [built_collection built_value http_multi_server logging path pool shelf shelf_web_socket stream_transform watcher web_socket_channel]
- build_resolvers 2.0.9 [analyzer async build crypto graphs logging path package_config pool pub_semver stream_transform yaml]
- build_runner_core 7.2.3 [async build build_config build_resolvers collection convert crypto glob graphs json_annotation logging meta path package_config pool timing watcher yaml]
- built_collection 5.1.1
- built_value 8.4.0 [built_collection collection fixnum meta]
- characters 1.2.0
- charcode 1.3.1
- checked_yaml 2.0.1 [json_annotation source_span yaml]
- cli_util 0.3.5 [meta path]
- clock 1.1.0
- cloud_firestore_platform_interface 5.7.3 [collection firebase_core flutter meta plugin_platform_interface]
- cloud_firestore_web 2.8.6 [cloud_firestore_platform_interface collection firebase_core firebase_core_web flutter flutter_web_plugins js]
- code_builder 4.1.0 [built_collection built_value collection matcher meta]
- collection 1.16.0
- convert 3.0.2 [typed_data]
- crypto 3.0.2 [typed_data]
- dart_style 2.2.3 [analyzer args path pub_semver source_span]
- device_info_plus_linux 3.0.0 [device_info_plus_platform_interface file flutter meta]
- device_info_plus_macos 3.0.0 [device_info_plus_platform_interface flutter]
- device_info_plus_platform_interface 3.0.0 [flutter meta plugin_platform_interface]
- device_info_plus_web 3.0.0 [device_info_plus_platform_interface flutter_web_plugins flutter]
- device_info_plus_windows 4.0.0 [device_info_plus_platform_interface ffi flutter win32]
- equatable 2.0.3 [collection meta]
- fake_async 1.3.0 [clock collection]
- ffi 2.0.1
- file 6.1.2 [meta path]
- firebase_analytics_platform_interface 3.3.3 [firebase_core flutter meta plugin_platform_interface]
- firebase_analytics_web 0.4.2+3 [firebase_analytics_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins js]
- firebase_auth_platform_interface 6.6.0 [collection firebase_core flutter meta plugin_platform_interface]
- firebase_auth_web 4.3.0 [firebase_auth_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins http_parser intl js meta]
- firebase_core_platform_interface 4.5.1 [collection flutter flutter_test meta plugin_platform_interface]
- firebase_core_web 1.7.2 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_crashlytics_platform_interface 3.2.15 [collection firebase_core flutter meta plugin_platform_interface]
- firebase_in_app_messaging_platform_interface 0.2.1+15 [firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_platform_interface 4.1.3 [firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_web 3.1.3 [firebase_core firebase_core_web firebase_messaging_platform_interface flutter flutter_web_plugins js meta]
- firebase_remote_config_platform_interface 1.1.15 [firebase_core flutter meta plugin_platform_interface]
- firebase_remote_config_web 1.1.4 [firebase_core firebase_core_web firebase_remote_config_platform_interface flutter flutter_web_plugins js]
- fixnum 1.0.1
- floor_annotation 1.3.0 [meta]
- flutter_keyboard_visibility_platform_interface 2.0.0 [flutter meta plugin_platform_interface]
- flutter_keyboard_visibility_web 2.0.0 [flutter_keyboard_visibility_platform_interface flutter_web_plugins flutter]
- flutter_plugin_android_lifecycle 2.0.7 [flutter]
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher material_color_utilities meta source_span stream_channel string_scanner term_glyph]
- flutter_web_plugins 0.0.0 [flutter js characters collection material_color_utilities meta vector_math]
- font_awesome_flutter 9.2.0 [flutter]
- frontend_server_client 2.1.3 [async path]
- glob 2.1.0 [async collection file path string_scanner]
- google_api_headers 1.3.0 [flutter package_info_plus]
- google_maps_flutter_android 2.1.10 [flutter flutter_plugin_android_lifecycle google_maps_flutter_platform_interface stream_transform]
- google_maps_flutter_ios 2.1.11 [flutter google_maps_flutter_platform_interface stream_transform]
- google_maps_flutter_platform_interface 2.2.1 [collection flutter plugin_platform_interface stream_transform]
- google_sign_in_android 6.0.1 [flutter google_sign_in_platform_interface]
- google_sign_in_ios 5.4.0 [flutter google_sign_in_platform_interface]
- google_sign_in_platform_interface 2.2.0 [flutter quiver]
- google_sign_in_web 0.10.2 [flutter flutter_web_plugins google_sign_in_platform_interface js]
- graphs 2.1.0 [collection]
- http 0.13.4 [async http_parser meta path]
- http_multi_server 3.2.1 [async]
- http_parser 4.0.1 [collection source_span string_scanner typed_data]
- image 3.2.0 [archive meta xml]
- in_app_purchase_android 0.2.3+1 [collection flutter in_app_purchase_platform_interface json_annotation]
- in_app_purchase_platform_interface 1.3.1 [flutter plugin_platform_interface]
- in_app_purchase_storekit 0.3.1 [collection flutter in_app_purchase_platform_interface json_annotation]
- in_app_review_platform_interface 2.0.3 [flutter url_launcher plugin_platform_interface platform]
- io 1.0.3 [meta path string_scanner]
- js 0.6.4
- libphonenumber_platform_interface 0.3.1 [flutter plugin_platform_interface]
- libphonenumber_plugin 0.2.3 [flutter flutter_web_plugins libphonenumber_platform_interface libphonenumber_web libphonenumber]
- libphonenumber_web 0.2.0+1 [flutter flutter_web_plugins js libphonenumber_platform_interface]
- linkify 4.1.0
- lint 1.10.0
- lists 1.0.1 [meta]
- logging 1.0.2
- matcher 0.12.11 [stack_trace]
- material_color_utilities 0.1.4
- meta 1.7.0
- mime 1.0.2
- nested 1.0.0 [flutter]
- package_config 2.1.0 [path]
- package_info_plus_linux 1.0.5 [package_info_plus_platform_interface flutter path]
- package_info_plus_macos 1.3.0 [flutter]
- package_info_plus_platform_interface 1.0.2 [flutter meta plugin_platform_interface]
- package_info_plus_web 1.0.5 [flutter flutter_web_plugins http meta package_info_plus_platform_interface]
- package_info_plus_windows 2.0.0 [package_info_plus_platform_interface ffi flutter win32]
- path 1.8.1
- path_drawing 1.0.1 [vector_math meta path_parsing flutter]
- path_parsing 1.0.1 [vector_math meta]
- path_provider_linux 2.1.7 [ffi flutter path path_provider_platform_interface xdg_directories]
- path_provider_platform_interface 2.0.4 [flutter platform plugin_platform_interface]
- path_provider_windows 2.1.0 [ffi flutter path path_provider_platform_interface win32]
- permission_handler_android 10.0.0 [flutter permission_handler_platform_interface]
- permission_handler_apple 9.0.4 [flutter permission_handler_platform_interface]
- permission_handler_platform_interface 3.7.0 [flutter meta plugin_platform_interface]
- permission_handler_windows 0.1.0 [flutter permission_handler_platform_interface]
- petitparser 5.0.0 [meta]
- platform 3.1.0
- plugin_platform_interface 2.1.2 [meta]
- pool 1.5.1 [async stack_trace]
- process 4.2.4 [file path platform]
- pub_semver 2.1.1 [collection meta]
- pubspec_parse 1.2.0 [checked_yaml collection json_annotation pub_semver yaml]
- quick_actions_android 0.6.1 [flutter quick_actions_platform_interface]
- quick_actions_ios 0.6.0+11 [flutter quick_actions_platform_interface]
- quick_actions_platform_interface 1.0.2 [flutter plugin_platform_interface]
- quiver 3.1.0 [matcher]
- rxdart 0.26.0
- shared_preferences_android 2.0.12 [flutter shared_preferences_platform_interface]
- shared_preferences_ios 2.1.1 [flutter shared_preferences_platform_interface]
- shared_preferences_linux 2.1.1 [file flutter path path_provider_linux path_provider_platform_interface shared_preferences_platform_interface]
- shared_preferences_macos 2.0.4 [flutter shared_preferences_platform_interface]
- shared_preferences_platform_interface 2.0.0 [flutter]
- shared_preferences_web 2.0.4 [flutter flutter_web_plugins shared_preferences_platform_interface]
- shared_preferences_windows 2.1.1 [file flutter path path_provider_platform_interface path_provider_windows shared_preferences_platform_interface]
- shelf 1.3.1 [async collection http_parser path stack_trace stream_channel]
- shelf_web_socket 1.0.2 [shelf stream_channel web_socket_channel]
- sign_in_with_apple_platform_interface 1.0.0 [flutter plugin_platform_interface meta]
- sign_in_with_apple_web 1.0.1 [flutter flutter_web_plugins sign_in_with_apple_platform_interface js]
- sky_engine 0.0.99
- source_gen 1.2.2 [analyzer async build dart_style glob meta path source_span yaml]
- source_helper 1.3.2 [analyzer collection source_gen]
- source_span 1.8.2 [collection path term_glyph]
- sqflite_common 2.2.1+1 [synchronized path meta]
- sqflite_common_ffi 2.1.1+1 [sqlite3 sqflite_common synchronized path meta]
- sqlite3 1.7.2 [collection ffi js meta path]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- stream_transform 2.0.0
- string_scanner 1.1.0 [charcode source_span]
- strings 0.2.2 [characters unicode]
- synchronized 3.0.0+2
- term_glyph 1.2.0
- test_api 0.4.9 [async boolean_selector collection meta source_span stack_trace stream_channel string_scanner term_glyph matcher]
- timing 1.0.0 [json_annotation]
- typed_data 1.3.1 [collection]
- unicode 0.3.1 [lists]
- universal_io 2.0.4 [collection crypto meta typed_data]
- url_launcher_android 6.0.17 [flutter url_launcher_platform_interface]
- url_launcher_ios 6.0.17 [flutter url_launcher_platform_interface]
- url_launcher_linux 3.0.1 [flutter url_launcher_platform_interface]
- url_launcher_macos 3.0.1 [flutter url_launcher_platform_interface]
- url_launcher_platform_interface 2.1.0 [flutter plugin_platform_interface]
- url_launcher_web 2.0.12 [flutter flutter_web_plugins url_launcher_platform_interface]
- url_launcher_windows 3.0.1 [flutter url_launcher_platform_interface]
- vector_math 2.1.2
- visibility_detector 0.3.3 [flutter]
- watcher 1.0.1 [async path]
- web_socket_channel 2.2.0 [async crypto stream_channel]
- win32 2.7.0 [ffi]
- xdg_directories 0.2.0+1 [meta path process]
- xml 6.1.0 [collection meta petitparser]
- yaml 3.1.1 [collection source_span string_scanner]
```

</details>

---

**Comment by darshankawar** [comment]
Thanks for the report. I have not yet setup the new AppleProvider, so unable to verify this issue on Android.
Keeping it open and labeling for further insights.

Also, I see the links in the doc are throwing 404:

<img width="680" alt="Screenshot 2022-08-29 at 6 31 25 PM" src="https://user-images.githubusercontent.com/67046386/187209410-9ea295f5-db30-4f31-924f-2148a121b662.png">


/cc @Lyokone

**Comment by Lyokone** [comment]
@darshankawar The doc is supposed to be seen directly on the Firebase website: https://firebase.google.com/docs/auth/flutter/federated-auth  where it redirects properly

**Comment by jeanmatthieud** [comment]
I tested it on iOs, same issue. Not tested on Web.

I edited the issue to reflect that.

**Comment by Lyokone** [comment]
@jeanmatthieud Checking on the actual implementation of linkWithCredential on Android, the `AuthCredential` is a native object that cannot be sent back to Flutter. A native copy is kept.

There is no Token for the Apple sign in so you should keep an instance of the Flutter object and use it to linkWithCredential. The hashcode parameter is what we use to link back the instance on native.

**Comment by jeanmatthieud** [comment]
@Lyokone Could you please give me an example? I don't see any `AuthCredential` instance other than `UserCredential.credential` (as shown in my issue description).

I tried with `FirebaseAuth.instance.currentUser!.linkWithCredential(authResult.credentials)` but it doesn't work.
I get the following exception:
```
I/flutter (18605): ----------------FIREBASE CRASHLYTICS----------------
I/flutter (18605): [firebase_auth/invalid-credential] The supplied auth credential is malformed, has expired or is not currently supported.
I/flutter (18605): #0      _SignInPageState._genericLinkWithCredential
I/flutter (18605): <asynchronous suspension>
I/flutter (18605): #1      _SignInPageState._handleAppleSignIn
I/flutter (18605): <asynchronous suspension>
I/flutter (18605): #2      _SignInPageState.build.<anonymous closure>.<anonymous closure>
I/flutter (18605): <asynchronous suspension>
I/flutter (18605): ----------------------------------------------------
```

**Comment by Lyokone** [comment]
I'll check that, it might not be due to the empty "token" indeed.

**Comment by Lyokone** [comment]
After investigating the Native SDK I've opened an issue on Native SDK: https://github.com/firebase/firebase-android-sdk/issues/4064

**Comment by felixgabler** [comment]
Hi, thanks for investigating! I am not sure whether the issue you opened in the native SDK completely reflects what the issue is. It does not only affect anonymous users but any authentication method. The credential token will also be empty for email login. I think this gives the native issue even more priority because it is a bug in my opinion.

**Comment by Lyokone** [comment]
We were missing some function, that should be resolved once #9535 is merged.
